pkg/alertmanager/alertmanager_template_test.go

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package alertmanager
+
+import (
+	"net/url"
+	"testing"
+
+	"github.com/prometheus/alertmanager/template"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_withCustomFunctions(t *testing.T) {
+	type tc struct {
+		name        string
+		template    string
+		result      string
+		expectError bool
+	}
+	tmpl, err := template.FromGlobs([]string{}, withCustomFunctions("test"))
+	assert.NoError(t, err)
+	data := template.Data{}
+	cases := []tc{
+		{
+			name:     "template tenant ID",
+			template: "{{ tenantID }}",
+			result:   "test",
+		},
+		{
+			name:     "generate grafana explore URL",
+			template: `{{ grafanaExploreURL "https://foo.bar" "test_datasoruce" "now-12h" "now" "up{foo!=\"bar\"}" }}`,
+			result:   `https://foo.bar/explore?left=` + url.QueryEscape(`{"datasource":"test_datasoruce","range":{"from":"now-12h","to":"now"},"queries":[{"datasource":"test_datasoruce","expr":"up{foo!=\"bar\"}","instant":false,"range":true,"refId":"A"}]}`),
```

Comment:
I've tried to compare it with the URL generated by Grafana 9.4 and I got it slightly different (look at the `datasource` inside `queries`):

```
{
    "datasource":"P1809F7CD0C75ACF3",
    "queries":[{"refId":"A","datasource":{"type":"prometheus","uid":"P1809F7CD0C75ACF3"},"editorMode":"code","expr":"count(up)","legendFormat":"__auto","range":true,"instant":false}],
    "range":{"from":"now-1h","to":"now"}
}
```

Reply:
Hmm, yes, there are more somehow backward compatible ways to refer to datasource (even in the dashboard JSON IIRC).
But yes, you're right. I updated it and seems to be working even with our 9.2 Grafana instance, so hopefully will be compatible enough. 

PTAL https://github.com/grafana/mimir/pull/3849/commits/e7abe5cb4c9c324dba89ff53ad3e7f636143cf31